File: src/domain/ticket_kind.zig
//! Ticket Kind classifies Tickets as tasks or bugs.

const std = @import("std");

/// The category of a Ticket.
pub const TicketKind = enum {
    task,
    bug,

    /// Default Ticket Kind for `tk add` until `--bug` is implemented.
    pub const default: TicketKind = .task;

    /// SQLite storage and CLI rendering string.
    pub fn text(self: TicketKind) []const u8 {
        return switch (self) {
            .task => "task",
            .bug => "bug",
        };
    }

    /// Write the CLI rendering string for use with the `{f}` specifier.
    /// Single-sources the unstyled representation on `text()`; styled render
    /// sites still wrap `text()` through the Styler.
    pub fn format(self: TicketKind, writer: *std.Io.Writer) std.Io.Writer.Error!void {
        try writer.writeAll(self.text());
    }
};

test "TicketKind.format writes text() via {f}" {
    var buf: std.Io.Writer.Allocating = .init(std.testing.allocator);
    defer buf.deinit();
    try buf.writer.print("{f}", .{TicketKind.bug});
    try std.testing.expectEqualStrings("bug", buf.written());
}

File: crates/tk/src/domain/ticket_kind.rs
//! Ticket Kind classifies Tickets as tasks or bugs.
//!
//! The two-variant set is mirrored in
//! the V1 `items.ticket_kind` CHECK constraint; the `text()` spelling is the
//! storage contract.

use std::fmt;

/// The category of a Ticket. `TicketKind::Task` is the default for `tk add`
/// until `--bug` is implemented.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Default)]
pub enum TicketKind {
    #[default]
    Task,
    Bug,
}

impl TicketKind {
    /// SQLite storage spelling and the value the `items.ticket_kind` CHECK
    /// constraint accepts; also the lowercase form `tk list` renders.
    #[must_use]
    pub fn text(self) -> &'static str {
        match self {
            Self::Task => "task",
            Self::Bug => "bug",
        }
    }

    /// Capitalized noun for the `tk show` facet bar, mirroring
    /// [`ItemClass::label`](crate::domain::item_class::ItemClass::label) so the
    /// Ticket Kind and Item Class read in the same register (`Task`/`Bug`
    /// alongside `Epic`).
    #[must_use]
    pub fn label(self) -> &'static str {
        match self {
            Self::Task => "Task",
            Self::Bug => "Bug",
        }
    }
}

impl fmt::Display for TicketKind {
    /// Single-sources the unstyled representation on [`TicketKind::text`].
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(self.text())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn display_writes_text() {
        assert_eq!(format!("{}", TicketKind::Bug), "bug");
    }

    #[test]
    fn default_is_task() {
        assert_eq!(TicketKind::default(), TicketKind::Task);
    }

    #[test]
    fn label_is_capitalized_and_distinct_from_storage_text() {
        // The facet bar reads `Task`/`Bug`; the SQL CHECK column stays
        // lowercase. Guards against collapsing the two back together.
        assert_eq!(TicketKind::Task.label(), "Task");
        assert_eq!(TicketKind::Bug.label(), "Bug");
        assert_eq!(TicketKind::Bug.text(), "bug");
    }
}
